warehouse/attestations/_core.py

```diff
@@ -0,0 +1,105 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import hashlib
+import tempfile
+
+from pathlib import Path
+from typing import Literal
+
+from pydantic import BaseModel, TypeAdapter
+from pypi_attestations import Attestation
+
+from warehouse.packaging import File, ISimpleStorage
+
+
+class Publisher(BaseModel):
+    kind: str
+    """
+    The kind of Trusted Publisher.
+    """
+
+    claims: object
+    """
+    Claims specified by the publisher.
+    """
+
+
+class AttestationBundle(BaseModel):
+    publisher: Publisher
+    """
+    The publisher associated with this set of attestations.
+    """
+
+    attestations: list[Attestation]
+    """
+    The list of attestations included in this bundle.
+    """
+
+
+class Provenance(BaseModel):
+    version: Literal[1]
+    """
+    The provenance object's version, which is always 1.
+    """
+
+    attestation_bundles: list[AttestationBundle]
+    """
+    One or more attestation "bundles".
+    """
+
+
+def get_provenance_digest(request, file: File) -> str | None:
+    if not file.attestations:
+        return None
+
+    publisher_url = file.publisher_url
+    if not publisher_url:
+        return None  # TODO(dm)
+
+    provenance_file_path, provenance_digest = generate_provenance_file(
+        request, publisher_url, file
+    )
+    return provenance_digest
+
+
+def generate_provenance_file(
+    request, publisher_url: str, file: File
+) -> tuple[Path, str]:
+
+    storage = request.find_service(ISimpleStorage)
+    publisher = Publisher(kind=publisher_url, claims={})  # TODO(dm)
```

Comment:
> I think we currently don't store any trusted publisher claims.

We do 🙂 -- they're stored on the DB macaroon that the publisher is associated with. In this setting, they should be retrievable via `request.identity.claims`, since `identity` is a `PublisherTokenContext`:

https://github.com/pypi/warehouse/blob/c9dc901237d1f9d15abb777dc1a347cc09959b7d/warehouse/oidc/utils.py#L96-L119

Reply:
And here's where each publisher indicates which claims to store during minting:

https://github.com/pypi/warehouse/blob/c9dc901237d1f9d15abb777dc1a347cc09959b7d/warehouse/oidc/views.py#L275

Reply:
(So yes, we should retrieve these stored claims and include them in the provenance.)

Reply:
Ah but I think that is in the context of a file upload, right? During a file upload, the request contains that identity (if uploading via Trusted Publishing). But this is the code that generates the provenance file, which (in this implementation) happens when the user requests a release using the simple API

Reply:
Ah, right -- yeah, this is another reason why we might want to push provenance generation onto the publishing side 🙂